docs/architecture/adr-020-protobuf-transaction-encoding.md

```diff
@@ -24,76 +25,186 @@ With this in mind, we will tackle the migration path via two main areas, txs and
+}
+
+message SignerInfo {
+    PublicKey pub_key = 1;
+    SignMode mode = 2;
+}
+
+enum SignMode {
+    SIGN_MODE_DEFAULT = 0;
+    SIGN_MODE_LEGACY_AMINO = -1;
+    SIGN_MODE_EXTENDED = 1;
+}
+```
+
+As will be discussed below, in order to include as much of the `Tx` as possible
+in the `SignDoc`, `SignerInfo` is separated from signatures so that only the
+raw signatures themselves live outside of `TxBody`.
 
-The application will define a single canonical `Message` Protobuf message
-with a single `oneof` that implements the SDK's `Msg` interface.
+Because we are aiming for be a flexible, extensible cross-chain transaction
+format, new transaction processing options should be added to `TxBody` as soon
+those use cases are discovered, even if they can't be implemented yet.
 
-Example:
+Because there is coordination overhead in this, `TxBody` includes an
+`extension_options` field which can be used for any transaction processing
+options that are not already covered. App developers should, nevertheless,
+attempt to upstream important improvements to `Tx`.
 
-```protobuf
-// app/codec/codec.proto
+### Signing
 
-message Message {
-  option (cosmos_proto.interface_type) = "github.com/cosmos/cosmos-sdk/types.Msg";
+Signatures are structured using the `SignDoc` below which reuses `TxBody` and only
+adds the fields which are needed for signatures but not present on `TxBody`:
 
-  oneof sum {
-    cosmos_sdk.x.bank.v1.MsgSend              msg_send             = 1;
-    cosmos_sdk.x.bank.v1.MsgMultiSend         msg_multi_send       = 2;
-    cosmos_sdk.x.crisis.v1.MsgVerifyInvariant msg_verify_invariant = 3;
-    // ...
-  }
+```proto
+// types/types.proto
+message SignDoc {
+    TxBody body = 1;
+    string chain_id = 2;
+    uint64 account_number = 3;
+    uint64 account_sequence = 4;
 }
 ```
 
-Because an application needs to define it's unique `Message` Protobuf message, it
-will by proxy have to define a `Transaction` Protobuf message that encapsulates this
-`Message` type. The `Transaction` message type must implement the SDK's `Tx` interface.
+#### `SIGN_MODE_DEFAULT`
 
-Example:
+The default signing behavior is to sign the raw `TxBody` bytes as broadcast over
+the wire. This has the advantages of:
 
-```protobuf
-// app/codec/codec.proto
+* requiring the minimum additional client capabilities beyond a standard protocol
+buffers implementation
+* leaving effectively zero holes for transaction malleability (i.e. there are no
+subtle differences between the signing and encoding formats which could 
+potentially be exploited by an attacker)
 
-message Transaction {
-  cosmos_sdk.x.auth.v1.StdTxBase base = 1;
-  repeated Message               msgs = 2;
+In order to sign in the default mode, clients take the following steps:
+
+1. Encode `TxBody`
+
+2. Sign `SignDocRaw`
+
+The raw encoded `TxBody` bytes are encoded into `SignDocRaw` below so that the
+encoded body bytes exactly match the signed body bytes with no need for
```

Comment:
This reads a bit weird. You're not signing the body bytes, you're signing the sign doc, right?

Reply:
Yes, but you're just copying the bytes you already encoded for `TxBody` into `SignDocRaw`

Reply:
Right, it just reads a bit weird, but I understand the statement.

Reply:
I can try to reword

Reply:
This was also a bit confusing to me when reading.